EIPS/eip-draft.md

```diff
@@ -0,0 +1,226 @@
+
+This EIP becomes particularly relevant with the adoption of EIP-8037, which increases `GAS_CODE_DEPOSIT` from 200 to 1,900 gas per byte. Under EIP-8037, deploying a 24kB contract would cost approximately 46.6M gas for code deposit alone, making the deduplication discount economically significant for applications that deploy identical bytecode multiple times.
+
+## Motivation
+
+Currently, deploying duplicate bytecode costs the same as deploying new bytecode, even though Ethereum clients don't store duplicated code in their databases. When the same bytecode is deployed multiple times, clients store only one copy and have multiple accounts point to the same code hash. Under EIP-8037's proposed gas costs, deploying a 24kB contract costs approximately 46.6M gas for code deposit alone (`1,900 × 24,576`). This charge is unfair for duplicate deployments where no additional storage is consumed.
+
+A naive "check if code exists in database" approach would break consensus because different nodes have different database contents due to mostly Sync-mode differences:
+- Full-sync nodes: Retain all historical code, including from reverted/reorged transactions
+- Snap-sync nodes: Only store code reachable from the current state trie
+
+Empirical analysis reveals that approximately 27,869 bytecodes existed in full-synced node databases with no live account pointing to them (as of the Cancun fork). A database lookup `CodeExists(hash)` would yield different results on different nodes, causing different gas costs and breaking consensus.
+
+This proposal solves the problem by making deduplication checks explicit and deterministic through access lists, ensuring all nodes compute identical gas costs regardless of their database state. (Notice here that even if fully-synced clients have more codes, there are no accounts whose codeHash actually is referencing them. Thus, users can't profit from such discounts which keeps consensus safe).
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Access List Extension
+
+EIP-2930 access list tuples are extended with an optional `checkCodeHash` boolean field:
+
+```json
+{
+  "address": "0x...",
+  "storageKeys": ["0x..."],
+  "checkCodeHash": true
+}
+```
+
+### Consensus semantics:
+
+- The `checkCodeHash` field is OPTIONAL. If omitted, it defaults to `false`.
+- Transactions with `checkCodeHash` fields are valid both pre-fork and post-fork.
+- Pre-fork: Nodes MUST ignore the `checkCodeHash` field and MUST NOT grant deduplication discounts.
```

Comment:
prior to the fork, that might cause a chain split between the nodes that have the update that supports the field and those who don't, if some node with the support for the field includes a tx with the field (whose content will be ignored, but that the others won't accept).

See https://go.dev/play/p/XaCF-q9ROEH for a reproduction of the problem that would affect geth.